---
title: A Biomarker Cutoff Optimization in TCGA HNSCC 
subtitle: Do you have a subtitle? If so, write it here
titlerunning: Short form of title (if too long for head)
authorrunning: Short form of author list if too long for running head
thanks: | 
    Grants or other notes about the article that should go on the front 
    page should be placed here. General acknowledgments should be placed at the
    end of the article.

authors: 
- name: Li-Hsing Chi 1
  address: Department of YYY, University of XXX
  email: abc@def
  
- name: Âuthóř 2
  address: Department of ZZZ, University of WWW
  email: djf@wef

keywords:
- HNSCC
- TCGA
- survival
- cutoff

#PACS: 
#- PAC1
#- superPAC
    
MSC:
- MSC code 1
- MSC code 2    

abstract:
  Gene or protein expression data are usually represented by metric or at least ordinal variables. In order to translate a continuous variable into a clinical decision, it is necessary to determine a cutoff point and to stratify patients into two groups each requiring a different kind of treatment. Currently, there is no standard method or standard software for biomarker cutoff determination. Therefore, we developed Cutoff Finder, a bundle of optimization and visualization methods for cutoff determination that is accessible online. While one of the methods for cutoff optimization is based solely on the distribution of the marker under investigation, other methods optimize the correlation of the dichotomization with respect to an outcome or survival variable. We illustrate the functionality of Cutoff Finder by the analysis of the gene expression of estrogen receptor (ER) and progesterone receptor (PgR) in breast cancer tissues. This distribution of these important markers is analyzed and correlated with immunohistologically determined ER status.

bibliography: bibliography.bib
output: rticles::springer_article
---

# Introduction {#intro}

Your text comes here. Separate text sections with

# Material and method {#sec:1}
<!-- # since the date [2019/05/13] after -->
[Survival analysis]
install linux on GCP with ssh connection
install R base and Rstudio server
install R packages on CRAN repository and Github
load(file="whole_genome.Rda") # the name list of protein coding genome
# below 2 lines: it needs to be modified for a function cal
LUAD_n <- length(whole_genome) #last one 20499: "ZZZ3"

introduce R scripts: TCGA_HNSCC_marginFree.R  TCGA_HNSCC_marginS.R  cutofFinder_func_HNSCC.R  main_marginS_free_HNSCC.R
https://jhoonline.biomedcentral.com/articles/10.1186/s13045-016-0372-0
Features selection:
pathologic_T => clinical_T of coln_osccT
in table 3 and table 4: featuresUni




by Joy Su
*P value <0.05 was considered statistically significant (Student’s t test for continuous variables and Pearson’s chi-square test for variables). SD represents the standard deviation. The tumor stage, tumor, lymph node, and distal metastasis status were classified according to the international system for staging HNSCC



[IHC]
Immunohistochemical staining

Three representative 1-mm-diameter cores from each tumor, taken from formalin-fixed paraffin-embedded tissues, were selected for morphology typical of the diagnosis. Assessable cores were obtained in 90 cases. The histopathological diagnoses of all samples were reviewed and confirmed by a pathologist, Michael Hsiao. IHC staining was performed on serial 5-μm-thick tissue sections cut from the tissue microarray (TMA) using an automated immunostainer (Ventana, Tucson, AZ, USA). Briefly, the sections were first dewaxed in a 60 °C oven, deparaffinized in xylene, and rehydrated in graded alcohol. Antigens were retrieved by heat-induced antigen retrieval for 30 min in Tris-EDTA buffer. The slides were stained with a polyclonal rabbit anti-human GLUT4 antibody (1:750, Epitomics, Cambridge, MA, USA). The sections were subsequently counterstained with hematoxylin, dehydrated, and mounted. The IHC staining intensity was scored by two pathologists as follows: no cytoplasmic staining or cytoplasmic staining in <10% of tumor cells was defined as score 0; faint/barely perceptible partial cytoplasmic staining in >10% of tumor cells was defined as score 1+; moderate cytoplasmic staining in >10% of tumor cells was defined as score 2+; and strong cytoplasmic staining in >10% of tumor cells was defined as score 3+. Scores of 0 and 1+ were defined as low GLUT4 expression, while scores of 2+ and 3+ were defined as high GLUT4 expression.

[Case selection]
Case selection from TCGA and TMU cohort

In total, 90 patients diagnosed with head and neck squamous cell carcinoma at the Taipei Medical University Hospital in Taiwan from 1991 to 2010 were included in this study. Patients who received preoperative chemotherapy or radiation therapy were excluded. Clinical information and pathology data were collected via a retrospective review of patient medical records. All cases were staged according to the 7th edition of the Cancer Staging Manual of the American Joint Committee on Cancer (AJCC), and the histological cancer type was classified according to the World Health Organization (WHO) 2004 classification guidelines. Follow-up data were available in all cases, and the longest clinical follow-up time was 190 months. Overall survival and disease-free survival were defined as the intervals from surgery to death caused by head and neck squamous cell carcinoma and recurrence or distant metastasis, respectively. The study was performed with the approval of the Institutional Review Board and with permission from the ethics committee of the institution involved (TMU-IRB 99049).

[Statistical analysis]

The nonparametric Mann–Whitney U test was used to analyze the statistical significance of results from three independent experiments. Statistical analyses were performed using SPSS (Statistical Package for the Social Sciences) 17.0 software (SPSS, Chicago, IL, USA). A paired t test was performed to compare the GLUT4 IHC expression levels in cancer tissues and in the corresponding normal adjacent tissues. The association between clinicopathological categorical variables and the GLUT4 IHC expression levels were analyzed by Pearson’s chi-square test. Estimates of the survival rates were calculated using the Kaplan–Meier method and compared using the log-rank test. The follow-up time was censored if the patient was lost during follow-up. Univariate and multivariate analyses were performed using Cox proportional hazards regression analysis with and without an adjustment for GLUT4 IHC expression level, tumor stage, lymph node stage, and recurrence status. For all analyses, a P value of <0.05 was considered significant.




Text with citations by @Galyardt14mmm, [@Mislevy06Cog].
1) Implementation under Rstudio server! run .R code
github as a version control
workflow: data access GDC portal, data cleaning, clinicopathologic features determination, extraction RNAseq for each gene, run100 for Cutoff Determination with P-value plot, ranking by fold change and significant P-value. wenn diagram => yield a table of candidate genes.
ROC curve
The optimal cutoff is defined as the point with the most
significant (log-rank test) split. Hazard ratios (HRs) including
95% confidence intervals are calculated.

2) manuscript in kniR.rmd r markdown or sweav to generate reproducible research article
3) surgical margin issue in survival analysis (even it is not a biological feature, which somehow is an artificial result, there is the most important effect on patient's prognosis)
4) Peter (a list with ranking) of mRNA (RNAseq) or LncRNA
cTNM alone (in KM analysis);
instead of pTNM, clinical features 不必多 (Joy's manuscript: smoking (±), betel nut)
n=500 case number 要多一些
The criteria of candidate genes: more 2500 genes (top 100); gene of non-coding RNA ?
IHC to convince and validate those genes
<30 published articles in pubmed
not related with known HNSCC genes

## Subsection title {#sec:2}

as required. Don't forget to give each section
and subsection a unique label (see Sect. \ref{sec:1}).

# Discussion {#sec:4}
=Dato 評估一下
Peter (a list with ranking) of mRNA (RNAseq) or Peter本來應該是希望你算 lncRNA 這應該就有獨特性.
你的方法跟..之間有和差異嗎? prognoscan, cutoff finder (http://molpath.charite.de/cutoff/docs/CutoffFinder_PLoSONE_2012.pdf
http://molpath.charite.de/cutoff/index.jsp
Human protein atlas: https://www.proteinatlas.org/humanpathology, protein atlas 有 TCGA最適跟中切
跟FireBrowse: http://firebrowse.org/# => TCGA 有預後跟臨床關係的相關性 其中有顯著意義的probe
你應該要評估一下 你在TCGA mRNA的部分是否有利基? Or LncRNA

=professor 陳志榮
biological meanings (surgical margin is an artificial factor)
=> margin (-) vs margin (+) 比較才公平

=Michael2
Dato 共同第一作者: 成大 n=44 microarray
Scientific Report; template from 把joy的paper拿來學
Rationale of PMM1: metabolic genes (yes, this is publishable) => PMM1 in HNSCC re-run  cutoff finder
Broader gene candidate (first 100)
gene signature from betel nut
GO, demographic
Tracy (Glioma genes)
IHC lung test slide
TFs 1,500 to 3,000 Non coding 10,000

#### Paragraph headings 

Use paragraph headings as needed.

\begin{align}
a^2+b^2=c^2
\end{align}





# References
