---
# Preparing your manuscript
# Scientific Data: article type as Data Descriptor %>%  https://www.nature.com/sdata/publish/submission-guidelines#sec-1
#*** http://osf.io/ Open Science Framework (data repository), OSF; ORCID,  connected to Github, Mendely => Texchi https://osf.io/4c5t6/?
PREPARING YOUR FILES
Follow the checklist below to ensure that your LaTeX files will properly convert to PDF.
  Files were generated using pdfTeX Version 3.1415926-2.4-1.40.13, TeXLive 2012 or earlier versions.
  Verify your images, graphs, drawings or tables are in any of these formats: gif, jpg, png, ps, eps, pdf
  All supplemental LaTeX files such as classes/styles, reference files, and bibliographies are referenced properly in the main document. Verify that they are referenced and uploaded with exactly same name and capitalization.
  Verify your images are in single, flattened layer.
  The commands to include your images in the LaTeX (.tex) main document do not begin with dots or slashes before the image name. For example, the correct command would be "\includegraphics[xxx]{fig1.eps}" instead of "\includegraphics[xxx]{./fig1.eps}". It is recommended that all of your LaTeX- related files are created and saved in a single folder.


#*** sharlatex and overleaf, sync with Github (vim): https://www.overleaf.com/project/5d1f4419460443357489e46f
#Overleaf:
#https://www.overleaf.com/project/5bfff217eedfa811462b1ad5
#https://www.overleaf.com/project/5bfff383eedfa811462b1bef (template of Scientific Reports)
#
# { The workflow to create such a document with R is:
# http://www.geo.uzh.ch/microsite/reproducible_research/post/rr-r-publication/
Create an .Rmd file (Markdown with R code blocks, called “code chunks”)
Write your report and include your data, code, analysis and text
Use the knitr package to combine text and the R scripts and render a markdown file from the RMarkdown
Convert the files with Pandoc (TexStudio; MacTex or Miktex) to generate HTML (can be self-contained), PDF, Word documents

Structure your project into the following steps:
-Data collection (raw data)
-Preprocessing (raw data to processed data)
-Analysis (processed data to results)
-Presentation (results presentation)

==
Submissions should be organized as follows:

Cover letter
Title page

Manuscript text:

Abstract
Keywords
Introduction
Materials and Methods
Results
Discussion
Acknowledgements
References
Figure/Table Legends
Figures/Tables
Supplementary Material
# }
title: Interrogating Biomarkers in HNSCC Survival Assessment
A Global Genome-wide Scan with Cutoff Mining for Emerging Biomarkers in Head and Neck Squamous Cell Carcinoma
A Global Genome-wide Scan by Data-driven approach to Identify the Prognostic Markers for Head and Neck Squamous Cell Carcinoma
Chi, L.-H. et al., Sci. Rep. 7, 9031 (2019, in submission). 
subtitle: A Optimizated Cutoff Finder for Biomarker Discovery for Head and Neck Squamous Cell Carcinoma
titlerunning: 
Analysis of RNA-Seq expression data using RStudio, and Bioconductor\cite{Loraine2015}.
authorrunning: Short form of author list if too long for running head
thanks: | 
    Grants or other notes about the article that should go on the front 
    page should be placed here. General acknowledgments should be placed at the
    end of the article.

authors: 
- name: Li-Hsing Chi 1
  address: Department of YYY, University of XXX
  email: texchi2@gmail.com
  affilication: The Ph.D. Program for Translational Medicine, College of Medical Science and Technology, Taipei Medical University and Academia Sinica
- name: Âuthóř 2
  address: Department of ZZZ, University of WWW
  email: djf@wef

keywords:
- HNSCC
- GDC
- TCGA
- ICGC
- survival
- cutoff
- biomarker
- reproducibility

#PACS: 
#- PAC1
#- superPAC
    
MSC:
- MSC code 1
- MSC code 2    

abstract: {#abstract}
  A gene expression–based prognostic markers for HNSCC, 442 samples were collected from 6 institutions and processed at 1 of 4 laboratories using a single, reproducible protocol, and data were organized by TCGA and were analyzed using a validation step by Tex.
  protein-coding DNA genes (20500, 2%) of Human genomes (Number of genes: 60483).
  Gene or protein expression data are usually represented by metric or at least ordinal variables. In order to translate a continuous variable into a clinical decision, it is necessary to determine a cutoff point and to stratify patients into two groups each requiring a different kind of treatment.
	Currently, there is no standard method or standard software for biomarker cutoff determination. Therefore, we developed the Cutoff Finder, a bundle of optimization and visualization methods for cutoff determination that is accessible online. 
	This protocol describes using R Markdown and RStudio, user-friendly tools for statistical analysis and reproducible research in bioinformatics, to analyze and document analysis of an RNA-Seq data set of TCGA cohort, which were suffered from HNSCC\cite{Loraine2015}, as well as ICGC cohort (https://dcc.icgc.org).
	While one of the methods for cutoff optimization is based solely on the distribution of the marker under investigation, other methods optimize the correlation of the dichotomization with respect to an outcome or survival variable. We illustrate the functionality of Cutoff Finder by the analysis of the gene expression of EGFR and VEGF in HNSCC cancer tissues. 
	This distribution of these important biomarkers is analyzed and correlated with immunohistologically determined EGFR status.


---


#Introduction {#intro}
project rationale: 人工決定 cutoff 需要天啟, it is difficult.
About 95% of all oral cancers occur in persons over 40 years of age.[https://doi.org/10.1136/bmj.h4238]
https://www.cancer.gov/types/head-and-neck/hp HNSCC
Types of Molecular Markers: Diagnostic Markers, Prognostic Markers or Predictive Markers.\cite{JNCCN2001} Prognostic markers have an association with some clinical outcomes, such as overall survival or recurrence-free survival, independent of the treatment rendered.
Currently, there is no global view of the expression profiles of the human cancer genes expression obtained from next-generation sequencing (NGS) technologies such as RNA sequencing (RNA-seq). The expression data is available from a large-scale studies like The Cancer Genome Atlas (TCGA: http://cancergenome.nih.gov/).
The Genomic Data Commons (GDC) maintains standardized genomic, clinical, and biospecimen data from National Cancer Institute (NCI) programs including The Cancer Genome Atlas (TCGA) and Therapeutically Applicable Research To Generate Effective Treatments (TARGET). We utilized GDCRNATools and firebrowseR, the R/Bioconductor packages, for downloading, organizing, and integrative analyzing RNA expression data (e.x. RNA-seq) and clinicopathological data in the GDC portal.
We retrieved RNA sequencing (RNA-seq) data together with clinical metadata corresponding to the 33 different human cancers that are available in TCGA (table S1). As a result, data were collected from 9666 individuals out of the 11,000 cancer patients included in the TCGA project from the Genomic Data Commons (GDC) Data Portal (https://gdc-portal.nci.nih.gov/).
Clinical validation assesses the strength of association between the assay results and the clinical outcome of interest, whether it is diagnostic, prognostic, or predictive. 
A large number of measures are used to assess these associations. Although measures of statistical signi cance are important, robust measures of the strength of association, such as receiver operator characteristic (ROC) curves, the area under the curve of ROC analysis, and the sensitivity, speci c- ity, and positive predictive value of assays, are criti- cal to understanding the strength of their clinical association and help bridge analysis to clinical util- ity (discussed later). These analyses address whether one can be sure the clinical state is positive if the test is positive (positive predictive power), and vice versa (negative predictive power).
https://www.ncbi.nlm.nih.gov/pubmed/29190934 (ROC)
(ROC) All codes and instructions can be found at https://osf.io/ef7na/#
=> 

Biomarker discovery is in hurry for immunotherapy of HNSCC.
TCGA survival analysis in global genome scanning manner. with margin free tissue coort is mandidatate. close, indeterminate, or positive margins.

*** significant! https://www.ebsco.com/e/zh-tw/blog/whats-so-significant-about-statistical-significance1?utm_medium=email&utm_source=med_cp_tw&utm_campaign=med_insights-health-2019-jun_20190629&mkt_tok=eyJpIjoiWmpRek9ETmxZekkwTnpSbSIsInQiOiJtc3VnRGVsOE85MGxtbk5PN256cXJXT2p2aXdITHgrOWhhbXFmMGxNdzFOOEtzdG45eUt1UVViZVRMcWVlNlhFQTVjUThDZnF0a2laS3FRUFwvY1BnMkdwZlh2NFhCdEFZQXRzTXFGWVhHRUZwNCtDMzJsY29LNHlTRWhXRllCWFoifQ%3D%3D

Optimzed cutoff finding (reproducible survival analysis: more than clicking, coping and pasting, which crossing 20500 whole genomes. We might get the same result in each run of R script)
 R packages http://www.r-project.org are used
Survival analysis in our cutoffinder_func engine employs the minimum P-value approach [2] to find the cutoff in continuous gene expression measurement for grouping patients. First, patients are ordered by expression value of a given gene. Next, patients are divided into two (high and low) expression groups at all potential cutoff (counted by person in 30% to 70% of a cohort), and the risk differences of the two groups are estimated by log-rank test. Then, optimal cutoff that gives the most pronounced P-value (P min) is selected programmatically in terms of yielding KM plot and Cox PH table for each gene.\cite{Mizuno2009a}\cite{Miller1982}Formula
type I error was conducted by multiple correlated testing [36–38]. Thus, P-value correction by Bonferroni/FDR is applied to control the error rate.
ε, 1 - ε] denote the range of the quantile considered to be cutpoints. PrognoScan uses ε = 0.3 to avoid small groupings from cutpoints of < 0.3 or > 0.7 quantile.
The aim of Reproducible Bioinformatics project is the creation of easy to use Bioinformatics workflows that fullfill the following roles (Sandve et al. PLoS Comp Biol. 2013): to provide Public Access to Scripts, Runs, and Results.
Rules: creating a portable project (.Rproj in Rstudio environment), using code/script (R\cite{TheRDevelopmentCoreTeam2003}) to automate data manipulation steps and analyses, connecting the results to dynamic text (R Markdown, knitr and LaTex), with version control (git) al scripts, figures and documents.
- RNAseq workflow
- miRNAseq workflow
- ChIPseq workflow
https://github.com/texchi2/docker4seq

# Material and method
<!-- # since the date [2019/05/13] after -->
*** [template of my manuscript] 範例 
1) https://www.hindawi.com/journals/bmri/2019/ 5450617/
Construction of the ceRNA Network 
HOTTIP Functions as a Key Candidate Biomarker in Head and Neck Squamous Cell Carcinoma by Integrated Bioinformatic Analysis

2) Deep Learning–Based Multi-Omics Integration Robustly Predicts Survival in Liver Cancer\cite{Chaudhary2018}
http://clincancerres.aacrjournals.org/content/24/6/1248
3) A pathology atlas of the human cancer transcriptome. Uhlen et al., Science 357, 660 (2017)
\cite{Uhlen2017} (human protein atlas, from https://www.proteinatlas.org/about/licence)


#set up a data science environment on Google Cloud Platform (GCP) {#sec:1}
it is also fine:
# Data Processing (cleaning) Work flow
{
Data management: dummy coding for category data (gender, age...)
Save the raw data.
Ensure that raw data are backed up in more than one location.
Create the data you wish to see in the world.
Create analysis-friendly data.
Record all the steps used to process data.
Anticipate the need to use multiple tables, and use a unique identifier for every record.
Submit data to a reputable DOI-issuing repository so that others can access and cite it.

}
#***TCGA data portal {#sec:2}
TCGA-Assembler was used to download RNA-seq data from TCGA data portal. Raw counts data with paired samples (tumor and normal) were extracted and analyzed using DEseq R package with default parameters: method1⁄4‘blind’, sharingMode1⁄4‘fit-only’, fitType1⁄4‘local’. DEseq normalization method has been reported to outperform other normalization methods. Fold changes, not absolute expression values, are displayed based on analysis described above. False discovery rates are not defined due to the low number of replicates for samples. This approach allows the user to determine the significance of differentially expressed genes on an individual basis.\cite{Wan2015}

# [FirebrowseR] data preparation

OPEN JOURNAL SYSTEMS
Journal Help
USER
Username	
Password	
 Remember me

FEEDS
RSS
Atom
NOTIFICATIONS
View
Subscribe
JOURNAL CONTENT
Search 

Search Scope 



Browse
By Issue
By Author
By Title
FONT SIZE
Make font size smallerMake font size defaultMake font size larger
Home | Mission | Instructions for Authors | Style Guide | Editorial Board | Contact | Register
Current Volume | Archives | Search
Home > Style Guide
Style Guide

In addition to the general instructions for manuscripts, we require all submissions to be formatted using LaTeX and the JSS style files (linked below along with the templates). It is recommended to keep the LaTeX code as simple as possible, i.e., avoid inclusion of packages/commands that are not necessary. We ask you to pay special attention to the frequently asked questions (FAQ).

To facilitate authoring manuscripts in JSS style, we provide several templates on this page. Specifically, a short demo article is available using either plain LaTeX or in combined R/LaTeX. For outlining the typical structure of a JSS article some brief text snippets are employed that have been inspired by doi:10.18637/jss.v027.i08 (Zeileis et al. 2008), discussing count data regression in R.

For authors of R software we recommend to write an R/LaTeX document (.Rnw) using Sweave or knitr. From this the LaTeX document (.tex) can be obtained by weaving/knitting along with the R replication code (.R) by tangling.

LaTeX
When using plain LaTeX, you may start from the following article template: jss-article-tex.zip (including all necessary style files). The source file article.tex can be edited in any text editor or dedicated LaTeX editor (including RStudio). To produce the manuscript pdfLaTeX should be used for which dedicated editors usually have a "Compile PDF" button or something similar. On the command line texi2pdf is convenient:

texi2pdf article.tex
The corresponding references are in the BibTeX bibliography ref.bib. The replication code article.R has to be prepared separately and should ideally include comments that make it easy to match it to the manuscript.

Code Snippets
Code snippet contributions are produced in the same way as articles. You only have to switch the from \documentclass[article]{jss} to \documentclass[codesnippet]{jss} in the first line of the LaTeX document.

Software and Book Reviews
For software and book reviews, LaTeX templates are provided in jss-review-tex.zip (including all necessary style files). The procedure to generate PDF output does not change from the above LaTeX section.

R/LaTeX via Sweave or knitr
For producing a JSS article with R/Latex, please start from the following template: jss-article-rnw.zip (including all necessary style files). The source files for the JSS contribution are the article.Rnw manuscript and the ref.bib bibliography in BibTeX format. To produce the PDF manuscript the article must first be weaved:

Sweave("article.Rnw")
yielding article.tex along with the graphics file (here article-visualization.pdf). This LaTeX file can be compiled in the "usual" way to the PDF using pdfLaTeX, i.e., using the shell, some LaTeX editor (see also above), or simply with R:

library("tools")
texi2pdf("article.tex")
An alternative route for RStudio users is to use the "Compile PDF" button directly for the article.Rnw. However, make sure not to use concordance mode: Tools > Global Options > Sweave > Uncheck: Always enable Rnw concordance.

The replication code can be obtained by tangling:

Stangle("article.Rnw")
yielding article.R. However, editing the comments of this .R file typically makes it more accessible to the readers/reviewers.

If knitr is used to prepare the article.Rnw file, the render_sweave() hook should be used to ensure JSS style code formatting. The weaving and tangling can then by done by knit() and purl(), respectively.

Frequently Asked Questions
What are the most important style guidelines in JSS?
What are the capitalization rules in JSS? What is title style and sentence style?
How to cite software?
How to cite R packages?
What are the different \cite, \citet, \citep commands about?
How should abbreviations be formatted?
How to format figure/table captions?
How should code be formatted in the manuscript?
What is a good pen-to-paper ratio for my graphics?
Some of my graphics files are very large, what should I do?
How should my R package reflect that a manuscript about it was published in JSS?
How can I turn my JSS paper into an R package vignette?
Which naming conventions are used for software, journal, and publisher names in JSS?
My LaTeX paper does not compile when there is JSS markup in section titles, what should I do?
Compiling my LaTeX paper fails with an error at \begin{document}, what went wrong?
Miscellaneous
What are the most important style guidelines in JSS?
The items below provide a quick style checklist for the most important style guidelines in JSS. More details can be found in the JSS style manual (jss.pdf) and in the other FAQ items.

The manuscript can be compiled by pdfLaTeX.
\proglang, \pkg and \code have been used for highlighting throughout the paper (including titles and references), except where explicitly escaped.
References are provided in a .bib BibTeX database and included in the text by \cite, \citep, \citet, etc.
Titles and headers are formatted as described in the JSS manual:
\title in title style,
all titles in the BibTeX file in title style.
\section, \subsection, etc. in sentence style,
annotations of figures/tables (including captions) in sentence style
Figures, tables and equations are marked with a \label and referred to by \ref, e.g., Figure~\ref{...}.
Software packes are \cite{}d properly.
What are the capitalization rules in JSS? What is title style and sentence style?
In English there are basically two styles of capitalization in titles, typically referred to as "sentence style" (or "sentence case") and "title style" (or "title case"). Although there are few strict rules, the subsequent set of guidelines should be helpful.

Sentence style: Only the first word in a title is capitalized, as is the first word after a colon or a hyphen. Of course, proper names remain in upper case. A simple example would be
A fancy topic: Implementation in Java

Title style: All principal words should be capitalized. This includes the first and last words of a headline, and all nouns, pronouns, adjectives, verbs, adverbs, and subordinating conjunctions (if, because, as, that, etc.). Do capitalize the first word after a colon. Articles (a, an, the), coordinating conjunctions (and, but, or, nor, for), and prepositions of any length, are to remain lowercased. However, if any of these are the first or last word of the headline, they should be capitalized. If you have an abbreviation in your headline that is normally lowercase, it should be left lowercase, particularly abbreviations for units of measure. Two part words separated by a hyphen should have both words capitalized. Examples:

Come Join Us for a Celebration
Caring for Your Houseplants
We're Getting Ready for an Early Spring
The Forecast for Summer: Hot!
As the Wind Blows
Spraying Schedule Posted on Office Memo Board
Remember to Observe All Parking Rules
What Are They Fighting For?
A New Record: 37-in. Snow Fall Accumulation
Tick-Tock: It's Daylight Savings Time Again!

How to cite software?
Many software packages tell their users how they want to be cited, i.e., there may be a pointer on the webpage or in the manual to a suitable publication (book, journal article, technical reports, etc.). If there is no recommended citation, please cite the corresponding manual or webpage. An example is given below for SAS/STAT 9.1.

@Manual{SAS-STAT,
  author  = {{\proglang{SAS} Institute Inc.}},
  title   = {\proglang{SAS/STAT} Software, Version~9.1},
  year    = {2003},
  address = {Cary, NC},
  url     = {https://www.sas.com/}
}
How to cite R packages?
Please check if there is an official citation for the package. If so, this can be seen on the associated CRAN webpage, i.e., https://CRAN.R-project.org/package=foo or queried in R (if the package is installed) via citation("foo"). If there is no citation, please use a CRAN style reference as exemplified for the rJava package below.

In any case, please make sure that the BibTeX is valid, that the title is in title style and that \proglang/\pkg/\code are used where appropriate.
cite R packages: https://www.jstatsoft.org/pages/view/style
@Manual{rJava,
  title  = {\pkg{rJava}: Low-Level \proglang{R} to \proglang{Java} Interface},
  author = {Simon Urbanek},
  year   = {2009},
  note   = {\proglang{R}~package version~0.8-1},
  url    = {https://CRAN.R-project.org/package=rJava},
}
manuscript and TCGA survival analysis Interpretation ##
head and neck squamous cell carcinoma (HNSC)
#https://www.biostars.org/p/153013/ Tutorial: Survival analysis of TCGA patients
#integrating gene expression (RNASeq) data Why would you want to do survival analysis based on gene expression data? Well, let's say you have a number of GDCRNATools;

#genes that you are interested in and they are differentially expressed between
#tumor and normal samples, it would be very powerful to show that alteration in
#gene expression correlates with worse survival or earlier tumor recurrence.
#https://groups.google.com/forum/m/#!msg/ucsc-cancer-genomics-browser/YvKnWZSsw1Q/3IAkkEMyFa4J
#from Mary Goldman and Jing Zhu, UCSC Cancer Browser
#https://genome-cancer.ucsc.edu/ 

https://github.com/IARCbioinfo/awesome-TCGA # list of all usefull TCGA tools
# or https://gdc.cancer.gov/access-data/gdc-community-tools, such as GDCtools
# FirebrowseR - Paper describing the R FirebrowseR package.
# GenomicDataCommons - Paper describing the R GenomicDataCommons package which might programmatically access GDC (Genomic Data Commons) Data Portal to retrieve TCGA data set.
# 

## [1.Directly] start from beginning  get Broad Institute GDAC: TCGA/Firhose data into R (Retrieve TCGA CDEs verbatim) #
# FirebrowseR - An R package to download directly the results of the analyses performed by Firehose in R.
#go to FireBrowse ( http://gdac.broadinstitute.org/ ):
# install.packages("devtools")
# library("devtools")
# devtools::install_github("mariodeng/FirebrowseR") # with more features (81): such as residual_tumor, vital_status, days_to_last_followup, "tobacco_exposure"
Data analysis
Normalization, testing, and false discovery rate estimation for RNA-sequencing data (FDR)\cite{Li2012}
\cite{Storey2003}
https://gdc.cancer.gov/about-data/data-harmonization-and-generation/genomic-data-harmonization-0#Overview
https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/
RNA sequencing (RNA-seq) data from head and neck squamous cell carcinoma (HNSCC) in The Cancer Genome Atlas (TCGA) datasets were obtained from the UCSC Cancer Genomics Browser (https://genome-cancer.ucsc.edu).
https://portal.gdc.cancer.gov/projects/TCGA-HNSC?primarySites_size=20
RNA-Seq reads that have been aligned to the GRCh38 build.
FPKM is A normalized expression value that takes into account each gene length and the number of reads mapped to all protein-coding genes


Quantitative variables were analyzed using Student’s t-test. Survival was analyzed using the log-rank test for two group comparisons. Values of P<0.05 (two-sided) were considered statistically significant. Statistical analyses were carried out using EZR version 1.32 (Saitama Medical Center, Jichi Medical University, Saitama, Japan)51.
to use a lookup table to manage datas: source("addNewData.r")
https://www.hindawi.com/journals/bmri/2019/5450617/
In the present study, the integrated RNA expression profiles of HNSCC patients along with their clinical outcomes were obtained from the TCGA database to construct a lncRNA–miRNA–mRNA ceRNA network for HNSCC. 
In 2011, Salmena et al. proposed a competing endogenous RNA (ceRNA) hypothesis that depicts a complicated posttranscriptional regulatory network, in which lncRNAs act as ‘sponges’, soaking up miRNAs by binding to miRNA response elements, thereby unsuppressing the target genes of the respective miRNAs.

# library(devtools, quietly=TRUE)
# source_gist("https://gist.github.com/dfalster/5589956")
# addNewData("dataNew.csv", myData, allowedVars)



#[Survival analysis] {#sec:3}
#https://data-flair.training/blogs/survival-analysis-in-r/
# ***Work flow of survival analysis
{
there is no RFS data on HNSCC cohort
cutoff finder, survival analysis => P-value
=> in .xlsl: we have
Table 2. The clinicopathological features of HNSCC cohort and ZZZ3 expression (chi square test).
(tableChi1 <- contingencyBin(contingencyTCGA())) (RNAseq vs features), 
v***by Chi-square test of association (P-value) or,
[***by R^2 (Spearman correlation coefficient, rho)
cor(x, y, method=“spearman”)]

Table 3: Cox's proportional hazards regression analysis, univariate and multipvariate => OS hazard ratio (HR), Cox P-value and its 95% confidence interval (CI)
(xlsx.addHeader(wb, sheet, value=paste("Table 3. Univariate/Multivariate Cox's proportional hazards regression analyses on OS time of", geneName, "gene expression in ", TCGA_cohort), level=5)
xlsx.addLineBreak(sheet, 1))

[table 4: RFS] ditto, omit
(xlsx.addHeader(wb, sheet, value=paste("Table 4. Univariate/Multivariate Cox's proportional hazards regression analyses on RFS time of", geneName, "gene expression in ", TCGA_cohort), level=5)
xlsx.addLineBreak(sheet, 1))

Figures: KM plot, cumulative KM P-value plot
Table of significant KM P-value
}
cor() in R:
"pearson"
"kendall"
"spearman"
-The Pearson correlation evaluates the linear relationship between two continuous variables. 
-Spearman rank-order correlation 比較複雜
monotonic, no constant rate


#Statistical analysis method
Cross validation, training set on n=150
rms, validate()
https://bmcmedgenomics.biomedcentral.com/articles/10.1186/1755-8794-4-31
Method:
Computations were carried out in C++ using a Borland 5 compiler and Optivec 5.0 vector and matrix libraries, and R version 2.6.1 (including R "stats" package for smooth.spline and isoreg functions). Gene expression data were obtained from the BRB ArrayTools Data Archive for Human Cancer Gene expression (url: http://linus.nci.nih.gov/BRB-ArrayTools.html), except for [20] data which was retrieved from the Broad Institute website (url: http://www.broad.mit.edu/cgi-bin/cancer/datasets.cgi). Data were normalized using MAS 5.0 and median centering each array using either R or BRB ArrayTools (developed by Dr. Richard Simon).

All statistical analysis were performed with R (https://www.r-project.org/, v3.0.1) and R packages. Normalization of Affymetrix microarray data were implemented with R package “limma”26. The survival analysis, cox univariate regression, cox multivariate regression and cox probability hazard model were carried out on R package “survival”30. Random forest variable hunting was performed with package “randomForestSRC”. The ROC curve was plotted with R package “pROC”31, and the nomogram were plotted with R package “rms”32.

A 1-year, 3-year, 5-year, and 10-year time-dependent receiver operating characteristic (ROC) curve analysis were performed using "survival-ROC" package in R software to assess the predictive value of the risk score for time-dependent outcomes [13, 14]. The ROC curve was used to evaluate the sensitivity and specificity of HNSCC-specific signatures for predicting overall survival of HNSCC patients.

Descriptive statistics were used to summarize patient and treatment characteristics. NLR was calculated as NLR=ANC/ALC and dNLR was calculated as dNLR=ANC/(WBC−ALC). Progression-free survival (PFS) was determined from the first cycle of treatment to disease progression documented by imaging, or death (event), or last follow-up (censored). Overall survival (OS) was calculated from the first cycle of treatment to death (event) or last follow-up (censored). Multivariate Cox proportional hazard models were used to investigate associations of NLR with survival, adjusted for baseline characteristics. Results were presented as hazard ratios (HR) with 95% confidence intervals (CIs).
All statistical analyses were performed using SPSS v21.0 (Beijing Stats Data CO. Ltd, Beijing China) and R 2.15.1 (http://www.r-project.org/). A two-sided P<0.05 was considered statistically significant.
Student’s t-test and the χ2 test were used for the comparison to baseline clinical characteristics and dietary information between the derivation data and validation data. In this study, noninvasive risk factors were defined as factors that could be measured without taking a blood sample. The variables included in the model were divided into classic risk factors and dietary predictors. Classic risk factors included age, gender, BMI, waist circumference, regular exercise, labor intensity, alcohol consumption, cigarette smoking, education, hypertension and a family history of diabetes. Dietary predictors included the 14 food groups in the FFQ and total calorie intake. Their adjusted odds ratio (OR) and 95% confidence intervals (95% CI) were estimated using a multivariate backward logistic regression analysis. The diabetes risk score was developed for the two multivariate logistic regression models: a classic noninvasive risk score and a diet-containing risk score, and points were assigned to each variable based on the magnitude of its regression coefficient [22]. In the diabetes risk score system, we transformed the unit of diet consumption from gram per day to liang per day, to facilitate its use in China. The dietary predictors except diary in the risk model were continuous in order to use the full information for dietary predictors. The risk score allows decimal value if the weight is not in round number. For the dairy products, because 24.4% and 29.4% of the total subjects in the derivation data and validation data were never consume milk and its products, we transformed the dairy and its products variable into binary variable in the risk model. A total diabetes risk score for each individual was calculated as the sum of the points for each variable.

The classic noninvasive risk model and the diet-containing risk model were compared with Akaike’s information criterion (AIC), area under curve (AUC), IDI, NRI and the Hosmer-Lemeshow goodness-of-fit test (HL test) for model performance, discrimination and predictive ability, and general calibration in derivation data and validation data separately. An AIC difference between two models of 10 or greater was considered to be significant with a lower AIC value indicating a better model performance [23]. AUC, IDI and NRI were calculated to evaluate the model discrimination and predictive ability [24]. To apply NRI in the logistic regression model, we adopted the approach proposed by Pencina et al [25]. The Hosmer-Lemeshow goodness-of-fit test (HL test) was used to examine how well the predicted prevalence matched the observed prevalence of type 2 diabetes, and P-values of 0.01 or greater from the test were considered to indicate a good calibration.
(*** P-value < 0.05 or <= 0.05)?

#ROC and crossvalidation
ROC for survival analysis
https://uwspace.uwaterloo.ca/bitstream/handle/10012/3974/thesis.pdf;jsessionid=8D60B97D732BB68D7C5DAC65A5DD90D8?sequence=1

***ROC and cross validation on KM survival and Cox propotional hazards model:
Gene Expression-Based Survival Prediction in Lung Adenocarcinoma: A Multi-Site, Blinded Validation Study
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2667337/
https://stats.stackexchange.com/questions/374835/what-is-the-equivalent-of-cross-validation-for-a-fitted-cox-proportional-hazards
ROC for survival analysis (thesis)
https://uwspace.uwaterloo.ca/bitstream/handle/10012/3974/thesis.pdf;jsessionid=8D60B97D732BB68D7C5DAC65A5DD90D8?sequence=1

https://stats.idre.ucla.edu/r/dae/logit-regression/ Logit Regression
The best cutoff was derived using Cutoff Finder software [15] based on an R routine which optimized the significance of the split between Kaplan-Meier survival curves measured by the log-rank test. Statistical analyses were performed with Rstudio server on GCE.
he medians of the independent data groups were compared by using nonparametric Mann-Whitney (2 groups) and Kruskal-Wallis test (>2 groups). Correlation between stromal and intra-tumoral Treg numbers was assess by Spearman’s rho test. The optimal cutoff points of the population were calculated by using the Cutoff finder web application [22]. Recurrence-free survival (RFS) and overall survival (OS) analyses were performed using Kaplan-Meier curves. Univariate and multivariate Cox regression models were applied to calculate hazard ratio, 95% confidence interval and significance, and to assess the independent contributions of each factor to the RFS and OS. P-values < 0.05 were considered statistically significant. All statistical analyses were performed by using the IBM SPSS Statistics 23 (IBM, Ehningen, Germany).
The patients in the training data set were divided into two groups (high- and low-risk) according to the cutoff value for risk using Cutoff Finder. Survival times were compared between the two groups using the Kaplan-Meier analysis and log-rank test.

# statistics by Joy Su
*P value <0.05 was considered statistically significant (Student’s t test for continuous variables and Pearson’s chi-square test for variables). SD represents the standard deviation. The tumor stage, tumor, lymph node, and distal metastasis status were classified according to the international system for staging HNSCC
P-value notes:
Significant codes:  0 ???***??? 0.001 ???**??? 0.01 ???*??? 0.05 ???.??? 0.1 ??? ??? 1
if <0.001 => mark as "***"

 The agreement between judgments of the enhancement patterns was assessed by the 2 radiologists and was evaluated using the kappa consistency test. Kappa values ≥0.81, 0.61–0.80, and ≤0.60 were considered to reflect excellent, good, and poor agreement, respectively. Additionally, log-rank analyses of Kaplan–Meier survival curves were performed to compare the PFS and OS of the cohort. Factors that were significant (P < 0.05) in univariate analysis were entered into multivariate survival analysis based on the Cox proportional hazard ratio (HR) model.
      # contingency P-value
# Kaplan-Meier curve + Log-rank test, + Cox proportional regression
    # textbook: David Kleinbaum???Survival analysis: A self-learning text
Chi-square => Fisher exact test issue: the small cell values in the contingency table (unbalanced data) => Exact tests are preferred see e.g this for SPSS or this from Kang and Ahn 2008 which found that for highly unbalanced 2x2 tables a conditional exact test (e.g. Fisher's exact) is preferred. a rule of thumb: at least one of the counts <= 5.
H0: the survival is independent of the expression level of the individual gene

(As the p-value 0.04828 is less than the .05 significance level, we do reject the null hypothesis that the survival is independent of the expression level of the individual gene => the survival is associated of the expression level of the gene.)
https://www.reddit.com/r/rstats/comments/4bnlor/getting_chisquared_approximation_may_be_incorrect/
http://www.r-tutor.com/elementary-statistics/goodness-fit/chi-squared-test-independence



# R code implementation on GCE  {#sec:4}
install linux on GCP with ssh connection
install R base and Rstudio server
install R packages on CRAN repository and Github
load(file="whole_genome.Rda") # the name list of protein coding genome
# below 2 lines: it needs to be modified for a function cal
LUAD_n <- length(whole_genome) #last one 20499: "ZZZ3", keep it LUAD_n as gene number
**margin issue
Insufficient surgical margins and extracapsular extension are main
risks for a reduced overall and recurrence free survival.\cite{Backes2017}
\cite{Sailer2017}

[Statistical analysis]
Differences between the groups were analyzed using
the Chi square test and Fisher exact test for categorical,
and the unpaired student’s t-test for continuous variables.
ANOVA and Tukey’s post-hoc was performed for analysis
of more than two groups. Survival rates and curves were
calculated and illustrated by the Kaplan-Meier method and
further analyzed by the log-rank. Variables that revealed
prognostic or effect modifying potential on the outcome
were subsequently evaluated by the proportional Cox
regression for forward selection. p-values <0.05 were
considered statistically significant. Statistical analysis was
done using SPSS (SPSS Inc., Chicago, IL).\cite{Backes2017}

說明
venn_marginSFP.tiff was generated
{
R4> HNSCC_OS_marginS_pvalueBonferroni_sorted$gene_id
 [1] "ZNF557"    "DKK1"      "ZNF266"    "CAMK2N1"  
 [5] "IL19"      "MYO1H"     "STC2"      "PGK1"     
 [9] "SURF4"     "FCGBP"     "LOC148709" "USP10"    
[13] "EVPLL"     "PNMA5"     "NDFIP1"    "FOXA2"    
[17] "GRAP"      "KIAA1683"  "ZNF846"    "ZNF20"    
[21] "CELSR3"    "SLC26A9"   "FAM3D"     "GPR15"    
[25] "KLRA1"     "NPB"       "TCP11"     "STIP1"

R4> HNSCC_OS_marginFree_pvalueBonferroni_sorted$gene_id
 [1] "SURF4"     "LOC148709" "DKK1"      "MYO1H"    
 [5] "NDFIP1"    "IL19"      "ZNF266"    "SET"      
 [9] "ANO2"      "ZNF557"    "EIF2AK1"   "USP10"    
[13] "MAGIX"     "LOC440944"

R4> HNSCC_OS_marginPlus_pvalueBonferroni_sorted$gene_id
[1] "SFT2D1"  "IL17REL" "CTLA4"   "ZFP2"    "ZNF101" 
[6] "PFN2" 

HNSCC_OS_marginSFP_pvalueBonferroni_KM_candidate_cox.Rda
}
***=> 28 vs 14 with intersection of 9 candidate genes; 
$`margin[-]:margin[+/-]` of intersection
[1] "SURF4"     "LOC148709" "DKK1"      "MYO1H"    
[5] "NDFIP1"    "IL19"      "ZNF266"    "ZNF557"   
[9] "USP10"
=> 但是只要 cohort 不同，candidate genes 就不會一樣啊
19 genes exclusively in _marginS_ cohort => GSEA analysis
14 genes from _marginFree_ cohort =>
GSEA analysis
比較 pathway 的差異(so what:-)
==

**16789 # SLC35E2 RNAseq has two measurement for each participants => there is two genes: SLC35E2A and SLC35E2B
whole_genome: updated as 20500 genes (SLC35E2 split as SLC35E2A and SLC35E2B)
clean and rearrange by geneID => whole_genome has 20500.

TCGA states: "For mRNA and microRNA expression data, we typically compute the relative expression of an individual gene and tumor to the gene's expression distribution in a reference population. That reference population is either all tumors that are diploid for the gene in question, or, when available, normal adjacent tissue." (It is not always clear what the cell of origin of a tumor is, so the mRNA expression in normal adjacent tissue can sometimes be misleading, which is why expression is sometimes compared within the set of tumors only).


#introduce R scripts: 
xTCGA_HNSCC_marginFree.R  
xTCGA_HNSCC_marginS.R
TCGA_HNSCC_marginSFP.R
cutofFinder_func_HNSCC.R  
main_marginSFP_HNSCC.R
{TCGA datasets
The TCGA expression data were downloaded from the University of Santa Cruz cancer browser – version 2015. firebrowseR
The 564 normalized RNA-seq v2 profiles were generated by IlluminaHiSeq. HPV status for 72 HPV+ and 243 HPV− samples was obtained from ref. [60].
}

analysis_export.R
and tobacco_exposure.R
https://jhoonline.biomedcentral.com/articles/10.1186/s13045-016-0372-0
Features selection:
pathologic_T => clinical_T of coln_osccT
in table 3 and table 4: featuresUni



#run mainA (marginS) [since 2019/06/12]
genome-wide scan for margin 0 or 1 cohort ##
TCGA_HNSCC_marginS.R
call function(survival_marginS) on TCGA_HNSCC_marginS.R, 

# Our core engine: cutofFinder_func_HNSCC.R, call, run100
{  cutoffReturn <- cutofFinder_func(geneName, marginTag) 
rename LUAD.mRNA.Exp., 139 load(file="~/R/HNSCC.clinical.RNAseq.Fire.Rda") # n=521, as clean6_oscc
}

generate HNSCC.clinico_mRNA.Fire for each gene
*** check >70% RNAseq of a cohort is available in particular gene; e.x. XKRY, 100% is NaN
check complete.cases of margin, H.score_T and OS_IND, data cleaning 
100 run:  # slice by cases, # 30 percentile, 70 percentile (cut from case 129 to case 298 in  427 cohort)
to find lowest P-value according to KM survival analysis at optimized cutoff

run survival analysis: OS by mysurv <- Surv(oscc$OS..months._from.biopsy, oscc$OS_IND==1)
 [survfit] - Kaplan-Meier curve, KM P-value, or KM plot of OS
contingency table to find significant clinicopathological features


[COXPH modelling] with OS, RFS; HNSCC features: pathologic to 

#R2Excel export the optimized P-values for each genes

#run mainB (marginFree) ditto
call function(survival_marginFree) on  TCGA_HNSCC_marginFree.R
core engine: cutofFinder_func_HNSCC.R, call, run100

#run mainC (marginPlus) diito
HNSCC_survivalAnalysis_marginFree_
start to run [2019/06/28]


## [Results of _marginS_] {#sec:5}
***Work flow of result analysis
{
Table 1: candidate genes by KM adjusted P-value and Z-score summary with its scatter plot
Table 2A and 2B and 2C: bad guy candidate genes => KM P-value, Cox P-value and HR (>=1.8) and its venn diagram
Table 3A and 3B and 3C: good guy candidate genes => KM P-value, Cox P-value and HR (<0.5) and its venn diagram
}
***how to z-score/Feature Scaling{\todo}
{ => ***  z-score => "Feature Scaling" 須要 正名之
Z-score standardization
Feature Scaling - normalization by (here: scaling to [0, 1] range); Rescaling data to have values between 0 and 1. This is usually called feature scaling. (min-max scaling)
As a rule of thumb I’d say: When in doubt, just standardize the data, it shouldn’t hurt.
https://sebastianraschka.com/Articles/2014_about_feature_scaling.html#about-standardization
convert the "number" frequency to z-scores ("Z" because the standard normal distribution is also known as the "Z distribution").
The Standard Normal Distribution is a specific instance of the Normal Distribution that has a mean of ‘0’ and a standard deviation of ‘1’.
the Z-distribution is a way of naming the Standard Normal distribution.
Z-score = ( x – µ ) / σ
Standard Deviation denoted by the symbol (σ) , the greek letter for sigma, is nothing but the square root of the Variance. Whereas Variance is average of the squared differences from the Mean.\cite{Abraham de Moivre, 1738} https://www.ztable.net

*** a candidate table (HNSCC_OS_marginS_pvalue1e_6_zscore0_6), with "number of significant P-values" from cutoff finding (becoming z-score: bigger, much more significant cutoff "sites" 愈多被發現的機會，隨便切在 30%~70% slicing 都是 significant，代表這個基因有比較高的機會是 truly prognostic impact)
Z-score:a normalized number (frequency) of OS P-values
=> number/OS_pvalue: from cutoff finder, the number (frequency) of OS P-values, which KM P-value < 0.05, in this gene;
# higher probability to be "found" significantly on a random selection of "cutoff" value in the tranditional manner.
(x)z-score calculation = [(value gene X in tumor Y)-(mean gene X in normal)]/(standard deviation X in normal)
A z-score for a sample indicates the number of standard deviations away from the mean of expression in the reference.  The formula is : 
(x)z = (expression in tumor sample - mean expression in reference sample) / standard deviation of expression in reference sample
Assumption for z-score test:
-Has a sample size above 30,
-Has an known population standard deviation.
(if not, t-score should be used instead)


# => a kind of local minimal or global minimal of curve fitting x(Levenberg-Marquardt Optimization)? to be classified by neural networks. Learning or training stage is done by applied The Levenberg-Marquardt back propagation algorithm „trainlm‟ is a network learning function that updates weight and bias values according to Levenberg-Marquardt optimization.
}

***how about false discovery rate (FDR)? (\cite{Storey2003})


#proceed since [2019/06/26] in detail
post1 process table1 (KM): candidate_sample(KM) 
楊老師: Bonferroni correction (adjustment) sets the significance cut-off at α/n. of KM P-value; Cox P-value <=0.05 as well.
***yes, we need Bonferroni (adjustment) sets the significance cut-off at α/n 
the cutoff for significance is an adjusted P-value < 0.05/n (Bonferroni’s Approach, where n is the total number of expressed mRNAs in HNSCC).\cite{Dingerdissen2018} 
it is highly confidential. Bonferroni correction 是寶藏 (by 王紋章老師 [2019/09/04]）


# subsetting the candidated genes table, according to the P-value <= 0.05 (no Bonferroni_cutoff) and Z-score >= 0.6
Table 1: candidate genes by KM adjusted P-value and Z-score summary with its scatter plot
#x ***after Bonferroni correction => n=26 in _marginS_
#x ***after Bonferroni correction => n=? in _marginFree_

post2 process: append table2 (new variable: KM + Cox): which yield n=6601 genes on the list.
candidate_cox

#Pickup all significant genes list -> HNSCC_OS_marginS_THREE_pvalue005 
***Excluded the HNSCC cancer driver genes list: data mining from Embase
(n=6313, KM P-value)
plot x2:
Cox's Harzard Ratios (univariate) plot [Rplot07_cox_uniHR.tiff]
Cox's Harzard Ratios (multivariate) plot [Rplot07_cox_multiHR.tiff]


# pickup1 n=71 (bad guy genes) from HNSCC_OS_marginS_THREE_pvalue005; 
# Cox HR (>=1.8) & (uni_P_value <= 0.05) & (multi_P_value <= 0.05)
(Bonferroni_cutoff = 5.31011e-06 is veary restricted in this scenario)
venn diagram for intersection: [Rplot09_venn_HR1.5.tiff]
Table 2A and 2B and 2C: bad guy candidate genes => KM P-value, Cox P-value and HR (>=1.8) and its venn diagram

Table 3A and 3B and 3C: good guy candidate genes => KM P-value, Cox P-value and HR (<0.5) and its venn diagram
# pickup2 n=47 (good guy genes) from HNSCC_OS_marginS_THREE_pvalue005
#* Cox HR <=0.5 (good_FC)  & (uni_P_value <= 0.05) & (multi_P_value <= 0.05)
(Bonferroni_cutoff = 5.31011e-06 is very restricted in this scenario)
venn diagram for intersection: [Rplot09_venn_HR0.5.tiff]

#
#Export r2excel and .Rda
The z-score summary plot(yes)
The Venn diagram (bad guy) and the candidate genes on HNSCC_OS_marginS_candidates_Venn.xlsx , which was exported successfully.
on table 2B, 2C and 3B, 3C: ex. "NEK6" has bad impact on clinical T when expression is high



## [Results of _marginFree_] {#sec:6}
since [2019/06/26]

## [Results of _marginPlus_] {#sec:7}
since [2019/06/26]

[as required. Don't forget to give each section
and subsection a unique label (see Sect. \ref{sec:1}).]

Pathway analysis by DAVID\cite{Huang2009a}
*** 8 candidate genes has Previous HGNC Symbol -> Current Gene Symbol -> convert all 47 genes to ENTREZID
# Previous HGNC Symbol -> Current Gene Symbol -> ENTREZID
updated_geneNameX <- c(
	"ATHL1",	"PGGHG", 80162,
	"CCDC64",	"BICDL1", 92558,
	"CNIH",	"CNIH1", 10175,
	"DPH3B",	"DPH3P1", 100132911,
	"FAM10A4",	"ST13P4", 145165,
	"FAM55C",	"NXPE3", 91775,
	"KAZ",	"KAZN", 23254,
	"NDNL2",	"NSMCE3", 56160
) => TCGA needs to be updated
# CNIH has CNIH1 CNIH2 CNIH3 CNIH4 family
#The Kappa Statistic is a chance corrected measure of agreement between two sets of categorized data. Kappa result ranges from 0 to 1. The higher the value of Kappa, the stronger the agreement. If Kappa = 1, then there is perfect agreement. If Kappa = 0, then there is no agreement.
#> PANTHER.db ####
# *** http://pantherdb.org
there is PANTHER PIE CHARTs
# => MYO10 in nicotine Nicotinic acetylcholine receptor signaling pathway => Myosin
# http://pantherdb.org/genes/gene.do?acc=HUMAN%7CHGNC%3D7593%7CUniProtKB%3DQ9HD67
# ***TFs pickup from panther (see .txt and figure)
***Nicotinic Acetylcholine Receptors are found on striated muscle at the neuromuscular junction and in the brain. They are one of the best understood receptors, often used as the model ion channel/receptor. The understanding of these receptors' actions by Fatt and Katz (1952; J Physiol 117:109-128) led to a basic understanding of neuronal transmission of information. When activated by acetylcholine released from nerves, these receptors depolarize the muscle cell and trigger muscle contraction. The best known agonist for these receptors is nicotine, but nicotinic acetylcholine receptors in the brain (those effected by smoking cigarettes) are still relatively poorly understood compared to nicotinic acetylcholine receptors at the neuromuscular junction.

=We next performed gene set enrichment analysis on the cell surface genes preferentially overexpressed in HPV+ or HPV− or both HNSCC subtypes. Cell surface genes in HPV− tumors were enriched for Nicotine pathway, consistent with the role of tobacco consumption in HPV− HNSCC. 
=> *** nicotine
\cite{Sivarajah2019}
tobacco_exposure feature was created on [2019/07/27]
# 60.65% is in high risk smoking group (HNSCC cohort n=521)
# 'Lifelong Non-Smoker', 1
# 'Current Smoker', 2
# 'Current Reformed Smoker for > 15 yrs', 3
# 'Current Refomed Smoker for < or = 15 yrs', 4
# 'Current Reformed Smoker Duration Not Specified', 5
# ==> [tobacco_exposure] a new binomial variable (categorizing) as risk of [high/low]:
#       level low by c(1, 3, 5)
#       level high by c(2, 4)
# in hnscc_nicotine, n=521
# [11] "tobacco_exposure"             
1   2   3   4   5 
117 177  73 139   2 


=> submit DAVID
1)Transcription regulation, or TFs
Protein involved in the regulation of the transcription process.
2)Pathway: Proteoglycans in cancer => KEGG: https://david.ncifcrf.gov/kegg.jsp?path=hsa05205$Proteoglycans%20in%20cancer&termId=550028858&source=kegg

we conducted KEGG pathway analysis to pinpoint the pathways enriched in the two subtypes. As we have used EASE score in DAVID as the enrichment method, these results should be interpretive only.
Gene functional classification
The tool classified the example gene list (∼400 genes) into ten functional groups in an easily readable tabular format. An example output is illustrated in Figure 2 as well as in slide 5 of Supplementary Data 5. Gene groups (with significant enrichment scores ≥1), such as cytokines/chemokines (group 1: 3.39), kinases (group 2: 2.21), clathrin membrane fusion genes (group 3: 1.86), transcription factors (group 6: 1.39) and so on, can easily be identified. All of these gene groups are highly relevant to an HIV study and are therefore expected biological results16. Organizing the large gene list into gene groups allows investigators to quickly focus on the overall major common biology associated with a gene group rather than one gene at a time, thereby avoiding dilution of focus during the analysis due to too many single genes. Furthermore, the '2D View' function associated with each group is able to display all related terms and genes in detail in one picture, to examine their interrelationships. For example, for the kinase group (group 2), a user who is not familiar with kinases may explore the terms of kinase activity, transferase activity, ATP-binding, nucleotide binding, protein metabolism, tyrosine specificity, serine/threonine specificity, regulation of G protein signaling, signal transduction and so on in one view at the same time (slide 6 of Supplementary Data 5). Therefore, we can quickly learn the biology for the kinase group, with the above-mentioned related terms in a single view and also identify the fine differences among them. For example, there are two G-protein-coupled receptor kinases, three protein tyrosine kinases and six kinases involved in cell surface receptor-linked signal transduction among the 23 kinases within the group. The fine details may be very important for pinpointing the key biology associated with a study.\cite{HuangNP2009} Nature Protocols: https://www.nature.com/articles/nprot.2008.211
use the total genes in the genome as a global reference background. They tend to give more significant P-values, as compared to the tools (e.g. Onto-Express) using a narrowed-down set of genes (e.g. genes only existing on a microarray) as a gene reference background. In addition, DAVID (\cite{Huang2009a}) tends to be more conservative by using genes existing on the array and found to be associated with terms in the corresponding annotation categories, as the gene reference background. Many tools further allow users to upload a customized gene list as a gene reference background (Supplementary Data 1).\cist{HuangNuc2009}

# Discussion
ranking tables by Z-score
zcut <- zcut # 0.8; Z distribution: mean=0, Max.: 2.3034
Z-score > 0.8 should be applied; since top 1 bad and good guy gene has Z-score 2.3, which implies Z-score might serve as a surrgated entity for the most prognostic impact (such as significnat KM P-value as well as extreme Cox hazard ratios) 

ROC curve: install.packages("pROC"){\todo}
The optimal cutoff is defined as the point with the most
significant (log-rank test) split. Hazard ratios (HRs) including
95% confidence intervals are calculated.
*** the receiver operating characteristics (ROC) curves of single risk factor for diagnosing bone metastases in patients with HNSCC cancer.

AUC: https://dasanlin888.pixnet.net/blog/post/465238196-預測模型使用auc的限制~~晨晰統計林星帆顧問
biomarker: A加上baseline risk score之後，預測有增值效果（incremental value）; discrimination 回答以上四個問題時，常常以Receiver Operating Characteristic（ROC）的曲線下面積（Area under the curve, AUC）作為主要的統計方法。例如A與B的AUC分別是88%跟85%，此時可以使用DeLong test檢定兩個曲線下面積的差異是否達統計顯著，倘若A的面積顯著大於B（P < 0.05），那麼作者則宣稱A的預測表現優於B。

然而讀者需要先瞭解到何謂AUC（又稱為C-index or C-statistics）
https://dasanlin888.pixnet.net/blog/post/465996605-預測模型的新指標－idi與nri（第1篇，共2篇）~
https://dasanlin888.pixnet.net/blog/post/466650818-如何評估預測模型的表現（第2篇，共3篇）-~
https://www.rdocumentation.org/packages/PredictABEL/versions/1.2-2/topics/reclassification

Table 2 討論，RNAseq expression vs patient's gender 有相關=> 強度為何? 相關的強度如何測: by ϕ係數 (phi coefficient), coefficient of contingency or Cramer’s V coefficient

Clinical utility of a biomarker refers to the ability of the assay to improve clinical decision-making and patient outcomes. The cluster of biomarkers we found clearly distinguishes outcomes between "positive" and "negative" groups, however there is no differential treatment available for those two groups of patient. 
PROGgene http://genomics.jefferson.edu/proggene/; cancer prognostic database

#*publication # Excluding the HNSCC cancer driver genes ##
***Michael: less than 30 published articles in pubmed
{checking: article number for each genes{\todo}
list of Cancer Gene Census (575 genes)
list of HNSCC genes from Embase (pubmed_hnscc_genes.csv, and a barplot: HNSCC_genes_embase5_barplot.tiff, >= 5 papers; such as EGFR, p16, CD44, TP53 and TNM)
}
[GDC data portal]
***(OK)GDC data portal: d/l Cancer Gene Census (575 genes)
We used R packages "rentrez" and "pubmed.mineR" to process the abstracts from PubMed: pubtator_function
[PubMed] search keyword
"((((marker[Abstract] OR biomarker[Abstract])) AND (HNSCC OR HNSC OR oral))) AND (prognostic[Title] OR prognosis[Title])", retmax=2000
[embase] search keyword: (prognosis:ti OR prognostic:ti OR 'tumor marker') AND cancer:ti AND 'head and neck squamous cell carcinoma'; excluding other cancer type via "disease filter"
asscess at [2019/07/20] 503 articles
=> expanded as
('cancer of unknown primary site'/dm OR 'esophageal adenocarcinoma'/dm OR 'esophageal squamous cell carcinoma'/dm OR 'esophagus cancer'/dm OR 'esophagus carcinoma'/dm OR 'head and neck cancer'/dm OR 'head and neck carcinoma'/dm OR 'head and neck squamous cell carcinoma'/dm OR 'head and neck tumor'/dm OR 'head cancer'/dm OR 'hypopharynx cancer'/dm OR 'hypopharynx carcinoma'/dm OR 'hypopharynx squamous cell carcinoma'/dm OR 'larynx cancer'/dm OR 'larynx carcinoma'/dm OR 'larynx squamous cell carcinoma'/dm OR 'mouth cancer'/dm OR 'mouth carcinoma'/dm OR 'mouth squamous cell carcinoma'/dm OR 'nasopharynx cancer'/dm OR 'nasopharynx carcinoma'/dm OR 'neck cancer'/dm OR 'oropharynx cancer'/dm OR 'oropharynx carcinoma'/dm OR 'oropharynx squamous cell carcinoma'/dm OR 'oropharynx tumor'/dm OR 'pharynx cancer'/dm OR 'salivary gland cancer'/dm OR 'thyroid cancer'/dm OR 'tongue cancer'/dm OR 'tonsil cancer'/dm) AND (prognostic:ti OR 'tumor marker' OR 'biological marker' OR prognosis)
AND 'mRNA expression level'
There is n= 785 HNSCC related genes (list in supplementary) pubmed_hnscc_genes.Rda and pubmed_hnscc_genes.csv


# 6388 -> n=6313 genes for following venn.


#BioXpress*.csv # data from https://hive.biochemistry.gwu.edu/cgi-bin/prd/bioxpress/servlet.cgi
***and TMU biobank TMU-biobank@tmu.edu.tw(Oral cavity= 39:25) T/N specific gene signature (交集 with my gene cluster)


#Immunotherapy: tumor-homing stem cells (EGF, VEGF cytokines)
tumor-specific antigen (CD4 CD8)

=Dato 評估一下 lncRNA{\todo}
gdcSurvivalAnalysis: Univariate survival analysis of multiple genes
In Jialab-UCR/GDCRNATools: GDCRNATools: an R/Bioconductor package for integrative analysis of lncRNA, mRNA, and miRNA data in GDC
Peter (a list with ranking) of mRNA (RNAseq) or Peter本來應該是希望你算 lncRNA 這應該就有獨特性.(The role of long non-coding RNA (lncRNA) expression in human head and neck squamous cell carcinoma (HNSCC) is still poorly understood. In this study, UCSD researchers aimed at establishing the onco-lncRNAome profiling of HNSCC and to identify lncRNAs correlating with prognosis and patient survival.)
eukaryotic long non-coding RNAs (lncRNAs)
https://github.com/Jialab-UCR/GDCRNATools
#TANRIC database: https://www.tanric.org; https://ibl.mdanderson.org/tanric/_design/basic/download.html; The Atlas of Noncoding RNAs in Cancer: It characterizes the expression profiles of lncRNAs in large patient cohorts of 20 cancer types including TCGA, CCLE and other independent datasets. (Reference: http://cancerres.aacrjournals.org/content/75/18/3728)
http://biophy.dzu.edu.cn/EVLncRNAs
http://www.bio-bigdata.com/lnc2cancer/(2.0, lncRNAs regulated by microRNA (miRNA), Transcription factor (TF), variant and methylation): HOTAIR, lnc-JPH1-7 in HNSCC. (experimentally validated)
www.oncolnc.org, OncoLnc lets you see survival correlations for 21 cancers at a time for mRNAs, miRNAs, and lncRNAs.
TCGA-SARC: sarcoma:(The project has controlled access data which requires dbGaP Access.) https://portal.gdc.cancer.gov/projects/TCGA-SARC
[Using TCGA HNSCC RNA-sequencing data from 426 HNSCC and 42 adjacent normal tissues, the researchers found 728 lncRNA transcripts significantly and differentially expressed in HNSCC. Among the 728 lncRNAs, 55 lncRNAs were significantly associated with poor prognosis, such as overall survival and/or disease-free survival.]https://www.lncrnablog.com/unraveling-the-oral-cancer-lncrnaome/
Long non-coding RNAs (lncRNAs) have important biological functions and can be used as prognostic biomarkers in cancer. To identify a lncRNA prognostic signature for head and neck squamous cell carcinoma (HNSCC).
GDCRNAtools to conduct advanced
analyses of RNA-seq and miRNA-seq data in GDC data portal for identification of
lncRNA-miRNA-mRNA competing triplets in cancer.
[LncRNAs as prognostic biomarkers in head and neck squamous cell carcinoma (HNSCC)]
Chaudhary R., Wang X., Cao B., Gimbrone N.T., Slebos R.J.C., Chung C.H.
Cancer Research 2019 79:13 Supplement 
Researchers at the Shanghai Jiao Tong University School of Medicine analysed RNA-seq data derived from the TANRIC database to identify a lncRNA prognostic signature model using the orthogonal partial least squares discrimination analysis (OPLS-DA) and 1.5-fold expression change criterion methods. The prognosis prediction model based on the lncRNA signatures and clinical parameters were evaluated using the 5-fold cross validation method.
A total of 84 out of 3199 lncRNAs were significantly associated with the survival of patients with HNSCC. Using the OPLS-DA and 1.5-fold change selection criterion, 5 lncRNAs (KTN1-AS1, LINC00460, GUSBP11, LINC00923 and RP5-894A10.6) were further selected. The prediction power of each combination of the 5 lncRNAs was evaluated through the receiver operating characteristic (ROC) curve and a three-lncRNA panel (KTN1-AS1, LINC00460 and RP5-894A10.6) achieved the highest prognostic prediction power in the cohort. The patients were categorized into high- and low-risk groups based on their three-lncRNA profiles. Patients with high-risk scores had worse overall survival than those with low risk scores in the cohort. Multivariable Cox regression analyses showed that the lncRNA signature and tumour grade were independent prognostic factors for patients with HNSCC.

加分 3D data visiualiztion
https://datascienceplus.com/tag/data-visualisation/

#*** [重要討論] (multiple testing 是發生在 cutoff_finder.R 100切)
0) cutoff finding #***
Tex: cut by 人頭[30...70]
Dato 你的方法跟..之間有和差異嗎? {\todo}
a.prognoscan\cite{Mizuno2009a} http://dna00.bio.kyutech.ac.jp/PrognoScan/PrognoScan.html
拿我的 gene list 放進去比比看
{Survival analysis in PrognoScan employs the minimum P-value approach [2] to find the cutpoint in continuous gene expression measurement for grouping patients.}
b.cutoff finder\cite{Budczies2012} (http://molpath.charite.de/cutoff/docs/CutoffFinder_PLoSONE_2012.pdf
http://molpath.charite.de/cutoff/index.jsp
utility and cited by
https://www.nature.com/articles/s41598-017-09333-8 \cite{Chen2017b}, and DAVID\cite{Huang2009a}
c.Findcut\cite{Chang2017a}
d.Human protein atlas: Human Pathology Atlas available from https://www.proteinatlas.org/humanpathology
A pathology atlas of the human cancer transcriptome. Uhlen et al., Science 357, 660 (2017)
\cite{Uhlen2017}
{
有 TCGA最適跟中切 (P values less than 0.001 in maximally separated, and median cut)
(unprognostic: Genes with a median expression less than FPKM 1 were lowly expressed) 會被排除
TCGA survival
https://www.proteinatlas.org/about/assays+annotation#tcga_survival
Based on the FPKM value of each gene, patients were classified into two expression groups and the correlation between expression level and patient survival was examined. The prognosis of each group of patients was examined by Kaplan-Meier survival estimators, and the survival outcomes of the two groups were compared by log-rank tests. Both median and maximally separated Kaplan-Meier plots are presented in the Human Protein Atlas, and genes with log rank P values less than 0.001 in maximally separated Kaplan-Meier analysis were defined as prognostic genes. If the group of patients with high expression of a selected prognostic gene has a higher observed event than expected event, it is an unfavourable prognostic gene; otherwise, it is a favourable prognostic gene. Genes with a median expression less than FPKM 1 were lowly expressed, and classified as unprognostic in the database even if they exhibited significant prognostic effect in survival analysis.

CORRELATION ANALYSIS AND PROGNOSTIC GENES (HNSCC)
=> favorable 450 genes
https://www.proteinatlas.org/api/search_download.php?search=prognostic%3Ahead%20and%20neck%20cancer%20favourable%20AND%20sort_by%3Aprognostic%20head%20and%20neck%20cancer&columns=g,gs,eg,gd,pe,rnacas,rnacad,rnacass,rnacasm,abrr,prognostic_head_and_neck_cancer&format=tsv
=> unfavorable 340 genes https://www.proteinatlas.org/api/search_download.php?search=prognostic%3Ahead%20and%20neck%20cancer%20unfavourable%20AND%20sort_by%3Aprognostic%20head%20and%20neck%20cancer&columns=g,gs,eg,gd,pe,rnacass,rnacasm,prognostic_head_and_neck_cancer&format=tsv
}

跟FireBrowse: http://firebrowse.org/# => TCGA 有預後跟臨床關係的相關性 其中有顯著意義的probe
你應該要評估一下 你在TCGA mRNA的部分是否有利基? Or LncRNA{\todo} => GDCRNAtools may do it.

***Text with citations by @Galyardt14mmm, [@Mislevy06Cog].
1) Implementation under Rstudio server! run .R code
github as a version control
https://icgc.org/icgc/cgp/63/509/70269: TCGA, GDC...
workflow: data access GDC portal, data cleaning, clinicopathologic features determination, extraction RNAseq for each gene, run100 for Cutoff Determination with P-value plot, ranking by fold change and significant P-value. wenn diagram => yield a table of candidate genes.
***Vtreat package for data preparation
https://github.com/WinVector/vtreat

2) manuscript in kniR.rmd r markdown or sweav to generate reproducible research article{\todo}
##*** RR>reproducible research resume:
#load(file=file.path(path_ZSWIM2, paste("HNSCC_OS", marginTag, "THREE_pvalue005.Rda", sep="")))
## <<<
e.x. cox proportional hazard ratio in AACSL 
{
??
}

3) ***surgical margin issue in survival analysis (even it is not a biological feature, which somehow is an artificial result, it is one of the most important impact on patient's prognosis){\todo}
A Proposal to Redefine Close Surgical Margins in Squamous Cell Carcinoma of the Oral Tongue.\cite{Zanoni2017}
=professor 陳志榮
biological meanings (surgical margin is an artificial factor)
=> margin (-) vs margin (+) 比較才公平? (run it since [2019/06/28] done) # i6-marginPlus (genes 太少，因為 n=1xx)
=> unfavorable genes of HNSCC 的確有差別 (so what?:-)
HNSCC_OS_marginSFP_pvalueBonferroni_KM_candidate_cox.Rda
$`margin[-]:margin[±]` of intersection
[1] "SURF4"     "LOC148709" "DKK1"      "MYO1H"    
[5] "NDFIP1"    "IL19"      "ZNF266"    "ZNF557"   
[9] "USP10"


4) Peter (a list with ranking) of mRNA (RNAseq) or LncRNA
cTNM alone (in KM analysis);
(OK) instead of pTNM, clinical features 不必多 (Joy's manuscript: smoking risk (high/low), betel nut)
(ok) n=512 case number 要多一些
The criteria of candidate genes: more 2500 genes (top 100)
gene of non-coding RNA lncRNA in HNSCC?

5) ***[hypothesis] to prove by bioinformatic way: Nicotinic acetylcholine receptor signaling pathway.
This is especially important for the one who has an extensive history of tobacco use.
# nicotine consumption has oral cancer prognostic impact as well as tumor initiation
1) KM plot: P-value = 0.0431
run n=414 cohort, survival analysis without RNAseq, and categorized by tobacco_exposure high/low, P-value = 0.0431[2019/08/01]
(KMplot_OS_tobacco_exposure.tiff)
(n=414, 1 observation deleted due to missingness)
414 名口腔癌患者中，吸菸 vs  不吸菸者，五年存活率分別為  40% vs 60%
(P-value < 0.05) Test for difference (log-rank test) between groups
術後第一年的差別不大，大約 90% 能過關，接下來第二、三、四、五年，差距就拉開了。
2) coxph: table 3 OS:
coxph(formula = Surv(osccCleanNA$OS..months._from.biopsy, osccCleanNA$OS_IN == 
    1) ~ +Gender + ageDx + clinical_T + clinical_N + clinical_M + 
    stage + margin + tobacco, data = osccCleanNA)
Gender2      
ageDx      * 
clinical_T * 
clinical_N   
clinical_M **
stage      . 
margin1    **
marginNA     
tobacco2   * 
tobaccoNA    
---
Signif. codes:  
0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
..table 3: univariate/multivariate cox's PH
=> multivariate
prognosis was significant impacted by age, clinical T, clinical M, margin as well as tobacco. with HR as 1.4, 2.2, 9.4, 1.7 and 1.5 respectively.(HR 1.5 算是強大的影響力)
clinical M and margin: P-value < 0.005
clinical T: HR 2.2 with 95CI as 1.2 ~ 4.2
tobacco and age: HR around 1.4 with 95CI as 1 ~ 2.
***=> age is following the time, while tobacco depends on your mind.
=> univariate
T (HR 1.4), M (4.1), margin (1.6), tobacco (1.4)
3) Clinical features of the patients with reactive and non-reactive lymphadenopathy were evaluated. We used Fisher’s exact test for categorical variables and the Student’s t-test or Wilcoxon test for continuous variables, as guided by the normality of the data. We calculated odds ratios (OR) for bivariate predictors of etiology. We performed multivariate logistic regression to determine independent predictors of lymphadenopathy etiology and summarized these results with adjusted odds ratios (AOR). CD4 cell count was log-transformed to normalize the distribution. A cutoff criteria of p<0.2 on bivariate analysis was used to select variables for imputation into the multivariate model. We performed Allen-Cady modified, backward selection procedure with age, sex, viral load detectability and CD4 cell count nadir included by default. Bivariate and multivariate analyses were carried out using R (R Foundation for Statistical Computing). \cite{Bogoch2013}



# ***tobacco article review...Embase or PubMed
*** tobacco exposure: 1 pack year
A way to measure the amount a person has smoked over a long period of time. It is calculated by multiplying the number of packs of cigarettes smoked per day by the number of years the person has smoked. For example, 1 pack year is equal to smoking 1 pack per day for 1 year, or 2 packs per day for half a year, and so on.
For example: a person who has smoked 15 cigarettes a day for 40 years has a (15/20) x 40 = 30 pack-year smoking history. One pack-year is smoking 20 cigarettes a day for one year. If someone has smoked 10 cigarettes a day for 6 years they would have a 3 pack-year history. accessed at 2019/08/09: https://www.cancer.gov/publications/dictionaries/cancer-terms/def/pack-year?redirect=true

{\cite{Mira2019}
The database of the Surveillance, Epidemiology, and End Results (SEER) analytical study reported that the primary cancers of the oral cavity, pharynx, and esophagus are strongly related to tobacco exposure and are also associated with alcohol abuse and low fruit intake. SEER-based research also found that the survivors of cancers of the oral cavity and pharynx have more than a two-fold excess risk of developing subsequent primary cancers (SPCs), and tobacco and/or alcohol consumption probably account for much of the increased risk [7, 14]
7.\cite{Weir2016}
14.Caporaso NE, Dodd KW, Tucker MA. New malignancies following cancer of the respiratory tract. In: Curtis RE, Freedman DM, Ron E, et al., eds. New Malignancies Among Cancer Survivors: SEER Cancer Registries, 1973-2000, 2006. Bethesda, MD p. 145-180. (NIH Publ. No. 05-5302; Available at http://seer.cancer.gov/publications).
https://seer.cancer.gov/data-software/
https://surveillance.cancer.gov/tools/
https://www.ncbi.nlm.nih.gov/pubmed/26286216
https://www.ncbi.nlm.nih.gov/pubmed/28743940
in R: SEERaBomb: https://github.com/radivot/SEERaBomb demo
http://epbi-radivot.cwru.edu/SEERaBomb/SEERaBomb.html
}
Five years after quitting, Your risk of cancers of the mouth, throat, esophagus, and bladder is cut in half. Cervical cancer risk falls to that of a non-smoker. By American Cancer Society https://www.cancer.org/healthy/stay-away-from-tobacco/benefits-of-quitting-smoking-over-time.html
Tobacco Control: Reversal of Risk after Quitting Smoking
IARC Handbooks of Cancer Prevention Volume 11
http://publications.iarc.fr/Book-And-Report-Series/Iarc-Handbooks-Of-Cancer-Prevention/Tobacco-Control-Reversal-Of-Risk-After-Quitting-Smoking-2007
The relative risk for oral and pharyngeal cancers remains higher for at least 20 years after cessation.\cite{Dresler2006} (Lancet)

the TP53 pathway (p14ARF, MDM2, p21, BAX and HMGA2): DOI 10.1007/s10654-016-0183-1
Analysis of gene expression networks in head and neck squamous cell carcinoma
European Journal of Epidemiology 2016 31 Supplement 1 (S55)
=> acting upstream (p14ARF, HMGA2, MDM2) -> TP53 -> downstream (p21(CDKN1A), BAX)
High p14ARF(CDKN2A) expression is a hallmark of human papilloma virus (HPV) associated HNSCC and patients have a favorable outcome.

The changes in TP53 transcriptional regulation, mammalian target of Rapamycin (mTOR) signaling and mitochondrial function were associated with a history of tobacco use. [Examining the contribution of smoking and HPV towards the etiology of oral cavity squamous cell carcinoma using high-throughput sequencing: A prospective observational study] \cite{Zammit2018}

Association between cigarette smoking and mutation of the p53 gene in squamous-cell carcinoma of the head and neck.
Brennan J Boyle J Koch W Goodman S Hruban R et. al.
The New England journal of medicine
1995 vol: 332 (11) pp: 712-7

**** TCGA:  (smoking/smokeless, alcohol, egfr_amplication_status, HPV) status is available ####
Epidermal growth factor receptor (EGFR) as a biomarker of smoking status and survival in oropharyngeal squamous cell carcinoma. 5-year disease specific survival (DSS). A Cox proportional hazard model of DSS showed smoking status and EGFR expression to be dependent of each other on predicting 5-year DSS. EGFR expression has been associated with several downstream pathways leading to a high tumor proliferation rate, inhibition of apoptosis, enhanced tumor invasion, and metastasis [6]. EGFR protein over-expression has been reported in 70 to 100% of head and neck squamous cell carcinomas (HNSCCs), and 46–72% of OPSCCs.\cite{Sivarajah2019}
The correlation between smoking and prognosis amongst HPV-positive and negative OPSCC patients has been well established in the literature [2, 3, 15]. Kumar et al. observed that EGFR expression was significantly higher in current smokers than in past smokers, who in turn had higher EGFR levels than those who never smoked [16]. This finding was corroborated by Baumeister et al. [6]. However, these studies did not define pack-years of tobacco smoking, include survival data.

6)
=Michael2
Dato 共同第一作者: 成大 n=44 microarray
Scientific Report; template from 把joy的paper拿來學
Rationale of PMM1: metabolic genes (yes, this is publishable) => PMM1 in HNSCC 
re-run  cutoff finder(done)
Broader gene candidate (first 100)
gene signature from betel nut{\todo}
GO, demographic
Tracy (Glioma genes)
IHC lung test slide
TFs 1,500 to 3,000 Non-coding lncRNA 10,000

{\todo}IHC to convince and validate those genes
less than 30 published articles in pubmed
not related with known HNSCC genes
# [IHC] Immunohistochemical staining {#sec:?}

Three representative 1-mm-diameter cores from each tumor, taken from formalin-fixed paraffin-embedded tissues, were selected for morphology typical of the diagnosis. Assessable cores were obtained in 90 cases. The histopathological diagnoses of all samples were reviewed and confirmed by a pathologist, Michael Hsiao. IHC staining was performed on serial 5-μm-thick tissue sections cut from the tissue microarray (TMA) using an automated immunostainer (Ventana, Tucson, AZ, USA). Briefly, the sections were first dewaxed in a 60 °C oven, deparaffinized in xylene, and rehydrated in graded alcohol. Antigens were retrieved by heat-induced antigen retrieval for 30 min in Tris-EDTA buffer. The slides were stained with a polyclonal rabbit anti-human GLUT4 antibody (1:750, Epitomics, Cambridge, MA, USA). The sections were subsequently counterstained with hematoxylin, dehydrated, and mounted. The IHC staining intensity was scored by two pathologists as follows: no cytoplasmic staining or cytoplasmic staining in <10% of tumor cells was defined as score 0; faint/barely perceptible partial cytoplasmic staining in >10% of tumor cells was defined as score 1+; moderate cytoplasmic staining in >10% of tumor cells was defined as score 2+; and strong cytoplasmic staining in >10% of tumor cells was defined as score 3+. Scores of 0 and 1+ were defined as low GLUT4 expression, while scores of 2+ and 3+ were defined as high GLUT4 expression.

[Case selection]
Case selection from TCGA and TMU cohort

In total, 90 patients diagnosed with head and neck squamous cell carcinoma at the Taipei Medical University Hospital in Taiwan from 1991 to 2010 were included in this study. Patients who received preoperative chemotherapy or radiation therapy were excluded. Clinical information and pathology data were collected via a retrospective review of patient medical records. All cases were staged according to the 7th edition of the Cancer Staging Manual of the American Joint Committee on Cancer (AJCC), and the histological cancer type was classified according to the World Health Organization (WHO) 2004 classification guidelines. Follow-up data were available in all cases, and the longest clinical follow-up time was 190 months. Overall survival and disease-free survival were defined as the intervals from surgery to death caused by head and neck squamous cell carcinoma and recurrence or distant metastasis, respectively. The study was performed with the approval of the Institutional Review Board and with permission from the ethics committee of the institution involved (TMU-IRB 99049).

***[Statistical analysis] for IHC

The nonparametric Mann–Whitney U test was used to analyze the statistical significance of results from three independent experiments. Statistical analyses were performed using SPSS (Statistical Package for the Social Sciences) 17.0 software (SPSS, Chicago, IL, USA). A paired t test was performed to compare the GLUT4 IHC expression levels in cancer tissues and in the corresponding normal adjacent tissues. The association between clinicopathological categorical variables and the GLUT4 IHC expression levels were analyzed by Pearson’s chi-square test. Estimates of the survival rates were calculated using the Kaplan–Meier method and compared using the log-rank test. The follow-up time was censored if the patient was lost during follow-up. Univariate and multivariate analyses were performed using Cox proportional hazards regression analysis with and without an adjustment for GLUT4 IHC expression level, tumor stage, lymph node stage, and recurrence status. For all analyses, a P value of <0.05 was considered significant.

#### Paragraph headings 

Use paragraph headings as needed.

\begin{align}
a^2+b^2=c^2
\end{align}

# Acknowledgements
The results shown here are in whole or part based upon data generated by the TCGA Research Network: https://www.cancer.gov/tcga.
The authors gratefully acknowledge contributions from the TCGA Research Network and its TCGA Pan-Cancer Analysis Working Group.
https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga/using-tcga/citing-tcga

# References{\todo}
bibliography: bibliography.bib
output: rticles::springer_article
